pkg/kv/kvserver/client_mvcc_gc_test.go

```diff
@@ -68,3 +89,198 @@ func TestMVCCGCCorrectStats(t *testing.T) {
+	insqlDB := execCfg.InternalDB
+	mkRecordAndProtect := func(ts hlc.Timestamp, target *ptpb.Target) *ptpb.Record {
+		recordID := uuid.MakeV4()
+		rec := jobsprotectedts.MakeRecord(recordID, int64(1), ts, nil, /* deprecatedSpans */
+			jobsprotectedts.Jobs, target)
+		require.NoError(t, insqlDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			return ptp.WithTxn(txn).Protect(ctx, rec)
+		}))
+		return rec
+	}
+
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+	runner.Exec(t, `CREATE TABLE foo (id INT PRIMARY KEY)`)
+
+	var tableID int
+	runner.QueryRow(t, "SELECT table_id FROM crdb_internal.tables WHERE name = 'foo'").Scan(&tableID)
+	require.NotEqual(t, 0, tableID)
+	tablePrefix := keys.SystemSQLCodec.TablePrefix(uint32(tableID))
+	tc.SplitRangeOrFatal(t, tablePrefix)
+	require.NoError(t, tc.WaitForSplitAndInitialization(tablePrefix))
+
+	// Lower the GC TTL of table foo and allow the zone config to get reconciled.
+	s := tc.Server(0)
+	store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
+	require.NoError(t, err)
+	repl := store.LookupReplica(roachpb.RKey(tablePrefix))
+	_, err = sqlDB.Exec(`ALTER TABLE foo CONFIGURE ZONE USING gc.ttlseconds = 1`)
+	require.NoError(t, err)
+	testutils.SucceedsSoon(t, func() error {
+		if repl.SpanConfig().GCPolicy.TTLSeconds != 1 {
+			return errors.New("waiting for span config to apply")
+		}
+		return nil
+	})
+
+	for i := 0; i < 10; i++ {
+		runner.Exec(t, `INSERT INTO foo VALUES ($1)`, i)
+	}
+
+	// Drop the table and wait for the MVCC GC schema change job to write its
+	// range deletion tombstone and start polling to check if the target span
+	// has been gc'ed.
+	runner.Exec(t, `DROP TABLE foo CASCADE `)
+	var jobID int64
+	runner.QueryRow(t, "SELECT job_id FROM [SHOW JOBS] WHERE job_type = 'SCHEMA CHANGE GC'").Scan(&jobID)
+	require.NotEqual(t, 0, jobID)
+	runner.CheckQueryResultsRetry(t, `
+SELECT count(*)
+  FROM (
+        SELECT job_id
+          FROM [SHOW JOBS]
+         WHERE job_type = 'SCHEMA CHANGE GC'
+               AND status = 'running'
+               AND running_status = 'waiting for MVCC GC'
+       )`,
+		[][]string{{"1"}})
+
+	// Write a protected timestamp record targeting the entire cluster's keyspace.
+	clusterTarget := ptpb.MakeClusterTarget()
+	protectedTime := timeutil.Now().UnixNano()
+	mkRecordAndProtect(hlc.Timestamp{WallTime: protectedTime}, clusterTarget)
+
+	// Wait for the protection policy to apply to foo's range.
+	testutils.SucceedsSoon(t, func() error {
+		if len(repl.SpanConfig().GCPolicy.ProtectionPolicies) == 0 {
+			return errors.New("waiting for span config to apply")
+		}
+		require.NoError(t, repl.ReadProtectedTimestampsForTesting(ctx))
+		return nil
+	})
+
+	// Put some garbage in the stats, so it triggers the mvcc gc queue.
+	ms := repl.GetMVCCStats()
+	ms.KeyBytes = 16 * (1 << 20) // 16mb
+	ms.ValBytes = 32 * (1 << 20) // 32mb
+	ms.GCBytesAge = 48 * (1 << 20) * 100 * int64(time.Hour.Seconds())
+	repl.SetMVCCStatsForTesting(&ms)
```

Comment:
Interesting that we needed this, in my own experiments I didn't but I was testing on master not 23.1.